File: Sources/OpenPocketViewCore/LocalVPNFilter.swift
import Foundation

/// Local VPN / ad-blocker filters (AdGuard, Blokada, RethinkDNS) capture the
/// camera UDP datalink. `bindProcessToNetwork` cannot bypass a `VpnService`
/// that did not call `allowBypass()`. Official camera apps have the same limit
/// (#239). Shells detect the tunnel; this type owns copy and when to hint.
public enum LocalVPNFilter: Sendable {
    /// First picture can take a few seconds (enable at +3 s; Pocket 3 format
    /// poke is later). Hint only after the well has sat black this long.
    public static let liveHintDelaySeconds: TimeInterval = 8

    /// Connection-setup banner when a local VPN is already on.
    public static let wizardBanner =
        "Pause VPNs and ad blockers, or exclude this app. They can block the camera live feed."

    /// WAITING FOR LIVE VIEW subtitle after [liveHintDelaySeconds] with no picture.
    public static let liveHint =
        "A VPN or ad blocker may be blocking the live feed. Pause it, or exclude this app, then try again."

    /// Join-Wi-Fi phone instruction. Same voice as the other wizard bullets.
    public static let joinWifiPhoneStep =
        "Pause VPNs and ad blockers, or exclude this app"

    public static func shouldHintOnLiveWait(
        vpnActive: Bool,
        hadVideo: Bool,
        secondsWithoutVideo: TimeInterval
    ) -> Bool {
        vpnActive && !hadVideo && secondsWithoutVideo >= liveHintDelaySeconds
    }

    /// CFNetwork `__SCOPED__` / ifaddrs names that mean a tunnel. `utun` is
    /// also used by iCloud Private Relay — shells should not warn on name
    /// alone before the live well has stalled.
    public static func isTunnelInterface(_ name: String) -> Bool {
        let lower = name.lowercased()
        return lower.contains("tap") || lower.contains("tun") || lower.contains("ppp")
            || lower.contains("ipsec")
    }
}

File: Apps/Android/app/src/main/kotlin/com/opencapture/openpocketcine/session/LocalVPNFilter.kt
package com.opencapture.openpocketcine.session

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import com.opencapture.openpocketcine.diagnostics.DiagnosticCenter

/**
 * Local VPN / ad-blocker filters (AdGuard, Blokada, RethinkDNS) capture the
 * camera UDP datalink. `bindProcessToNetwork` cannot bypass a [android.net.VpnService]
 * that did not call `allowBypass()`. Matches iOS `LocalVPNFilter` copy and
 * live-wait policy (#239).
 */
object LocalVPNFilter {
    const val LIVE_HINT_DELAY_SECONDS = 8.0
    const val LIVE_HINT_DELAY_MS = 8_000L

    const val WIZARD_BANNER =
        "Pause VPNs and ad blockers, or exclude this app. They can block the camera live feed."

    const val LIVE_HINT =
        "A VPN or ad blocker may be blocking the live feed. Pause it, or exclude this app, then try again."

    const val JOIN_WIFI_PHONE_STEP = "Pause VPNs and ad blockers, or exclude this app"

    fun shouldHintOnLiveWait(
        vpnActive: Boolean,
        hadVideo: Boolean,
        secondsWithoutVideo: Double,
    ): Boolean = vpnActive && !hadVideo && secondsWithoutVideo >= LIVE_HINT_DELAY_SECONDS

    fun isActive(context: Context): Boolean {
        val connectivity =
            context.applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE)
                as? ConnectivityManager ?: return false
        return isActive(connectivity)
    }

    @Suppress("DEPRECATION")
    fun isActive(connectivity: ConnectivityManager): Boolean =
        connectivity.allNetworks.any { network ->
            connectivity.getNetworkCapabilities(network)
                ?.hasTransport(NetworkCapabilities.TRANSPORT_VPN) == true
        }

    @Volatile private var loggedActive = false

    fun noteIfActive(context: Context) {
        if (!isActive(context) || loggedActive) return
        loggedActive = true
        DiagnosticCenter.log(
            "notice",
            "session",
            "vpn",
            "vpn: local VPN or ad blocker active — can drop UDP live view",
        )
    }
}
